x-pack/plugins/upgrade_assistant/public/application/components/overview/fix_logs_step/external_links.tsx

```diff
@@ -55,7 +55,14 @@ const DiscoverAppLink: FunctionComponent = () => {
         return;
       }
 
-      const url = await locator.getUrl({ indexPatternId });
+      const url = await locator.getUrl({
```

Comment:
I dug a bit into `useUrl` and its implemented as a hook, which we cannot call from within the useEffect we have here. I'm leaving the `getUrl` instead because it will be more straightforward to read compared to having another top level hook that depends on the `indexPatternId` generated by this `useEffect`.

Reply:
Perhaps in a separate PR we can change https://github.com/elastic/kibana/blob/master/src/plugins/share/common/url_service/locators/use_locator_url.ts#L42 to define the dependencies more comprehensively:

```js
  // Convert params object to string to be able to compare its value,
  // allowing the consumer to freely pass new objects to the hook on each
  // render without requiring it to be memoized.
  const paramsStringified = params ? JSON.stringify(params) : undefined;

  /* eslint-disable react-hooks/exhaustive-deps */
  useEffect(() => {
    if (!locator) {
      setUrl('');
      return;
    }

    locator
      .getUrl(params, getUrlParams)
      .then((result: string) => {
        if (!isMounted()) return;
        setUrl(result);
      })
      .catch((error) => {
        if (!isMounted()) return;
        // eslint-disable-next-line no-console
        console.error('useLocatorUrl', error);
        setUrl('');
      });
  }, [locator, paramsStringified, ...deps]);
  /* eslint-enable react-hooks/exhaustive-deps */
```

This would allow us to consume the hook, since it will update once the `indexPatternId` dependency has resolved:

```js
  /* 
   * All this code replaces the current `useEffect`
   */

  const [indexPatternId, setIndexPatternId] = useState<string | undefined>();

  useEffect(() => {
    async function fetchIndexPatternId() {
      setIndexPatternId(await getDeprecationIndexPatternId(dataService));
    }
    fetchIndexPatternId();
  }, []);

  const locator = share.url.locators.get('DISCOVER_APP_LOCATOR');
  const discoverUrl = locator?.useUrl({
    indexPatternId,
    query: {
      language: 'kuery',
      query: `@timestamp > "${lastCheckpoint}"`,
    },
  });
```

Reply:
👍🏼  I've created https://github.com/elastic/kibana/issues/111512 to track that